Support passing multiple values to spy

alysbrooks opened the pull request:
I often want to pass multiple values to spy, so I added that capability to the macro.

Having worked on this PR, I think the intended usage is wrapping existing expressions with `spy`, which avoids the need for passing multiple values to spy. Still, I think using spy with multiple values is sometimes useful even if you predominantly wrap existing forms.

alysbrooks committed:
Support passing multiple values to spy.

alysbrooks committed:
Fix issues with form expansion.

alysbrooks committed:
Add documentation and CHANGELOG entry.

alysbrooks committed:
Add better example and further explanation.

plexus commented:
Interesting. I don't really have a habit of using spy but I'm fine with this going in. Note though that spy is modelled after `io.pedestal.log/spy`, so I would also update `glogc` to support multiple expressions, which will mean reimplementing based on ~`io.pedestal.log/debug`~ `io.pedestal.log/log-expr`.

- https://github.com/lambdaisland/glogi/blob/main/src/lambdaisland/glogc.cljc#L54-L56
- http://pedestal.io/api/pedestal.log/io.pedestal.log.html#var-spy

plexus commented:
Took me a few minutes to understand your implementation, here's my attempt

```clojure
(defmacro spy
  ([form]
   (let [res (gensym)]
     `(let [~res ~form]
        ~(log-expr &form :debug [:spy `'~form
                                 :=> res])
        ~res)))
  ([form & forms]
   ;; using cons to make explicit that it's a prepend
   (let [forms (cons form forms)
         syms (repeatedly (count forms) gensym)
         ;; map/mapcat take multiple collections
         bindings (mapcat list syms forms)
         spy-vec (vec (mapcat (fn [form sym]
                                ;; or `'~form as above, but if you don't write a
                                ;; lot of macros that's a bit harder to parse
                                [(list 'quote form) sym])
                              forms
                              syms))]
     `(let [~@bindings]
        ~(log-expr &form :debug [:spy spy-vec])
        ~(last syms)))))
```

alysbrooks commented:
@plexus That is a lot more readable. I'll replace my implementation with yours.

alysbrooks committed:
Use Arne's clearer reimplmentation of spy.

Co-authored-by: Arne Brasseur <arne@arnebrasseur.net>

alysbrooks committed:
Add cljs macro to glogc.

alysbrooks committed:
Add some tests of glogc using glogi.

plexus commented on src/lambdaisland/glogc.cljc:
```diff
@@ -51,9 +51,13 @@
+        (target :clj `(pedestal/spy ~expr)
+                :cljs `(glogi/spy ~expr)))
+       ([expr & exprs]
+        (target :clj `(pedestal/spy ~expr)
```
I guess in the case of clj/pedestal we can `(pedestal/spy [~expr ~@exprs])`

alysbrooks replied:
Done!

plexus commented:
Looks good to me!

alysbrooks committed:
Fix bug and add workaround in glogc.

Accept Arne's suggestion to put expressions in a list.

It's not ideal, but passing values to spy in a cljc file will work, even
if the way it appears is harder to read.

Also fixes a bug where the macro would generate `glogi/spy` for Clojure,
rather than `pedestal/spy`.

Co-authored-by: Arne Brasseur <arne@arnebrasseur.net>

alysbrooks committed:
Update README to cover Clojure behavior difference.

alysbrooks committed:
Revise new docs around `spy`.

alysbrooks committed:
Make spy consistent across Clojure and ClojureScript.

Before, passing multiple values to spy in Clojure would return all
values in a vector. Now, it returns just the last value, just as it does
in ClojureScript.

alysbrooks commented on src/lambdaisland/glogc.cljc:
```diff
@@ -51,9 +51,13 @@
+        (target :clj `(pedestal/spy ~expr)
+                :cljs `(glogi/spy ~expr)))
+       ([expr & exprs]
+        (target :clj `(last (pedestal/spy [ ~expr ~@exprs]))
```
Had to add last here for consistency with ClojureScript. If `spy` were used for production logging, I would wonder  about this implicit vector, but it isn't.
